### ISSUE TITLE
[MVT] OL differences with Mapbox rendering

### ISSUE THREAD

**Issue by danduk82** [geoadmin/mf-geoadmin3#4752]
OL does not use the same restrictions on the zoomlevels.

The most visible issue are the mountains, but there is the same problem also with other features.

Just compare them [here](https://codepen.io/pakb/pen/oJNRyK)

Here is a list of issues opened in olms that are mandatory to solve before we can go live

- [x] https://github.com/openlayers/ol-mapbox-style/issues/136 Multiple source : Label decluterring
- [x] https://github.com/openlayers/ol-mapbox-style/issues/137 Add support for font styling described in mapbox style
- [x] https://github.com/openlayers/ol-mapbox-style/issues/138 Interpret zoom level in the same way mapbox does
- [x] https://github.com/openlayers/ol-mapbox-style/issues/139 Rendering of small features, min-width?
- [x] https://github.com/openlayers/ol-mapbox-style/pull/168 text-halo isn't large enough
- [x] https://github.com/openlayers/ol-mapbox-style/issues/170 label margin-bottom is not big enough, and labels tend to overlap their point
- [ ] https://github.com/openlayers/ol-mapbox-style/issues/185 line-offset not supported
- [x] https://github.com/openlayers/ol-mapbox-style/issues/186 labels following polylines bend a lot

**Comment by pakb** [comment]
I have a fix to push in https://github.com/geoadmin/ol-mapbox-style but I don't have the rights to write in this repo.
Could you change the permissions @danduk82 ?

**Comment by pakb** [comment]
Here's a codepen showing why we have a problem with zoom levels : https://codepen.io/pakb/full/NoLJYB

**Comment by pakb** [comment]
A fix is on the branch https://github.com/geoadmin/mf-geoadmin3/tree/feature/update_mb2olstyle and will be deployed on the testviewer in the next couple hours

**Comment by pakb** [comment]
Here is a list of issues opened in olms that are mandatory to solve before we can go live

- [ ] https://github.com/openlayers/ol-mapbox-style/issues/136 Multiple source : Label decluterring
- [ ] https://github.com/openlayers/ol-mapbox-style/issues/137 Add support for font styling described in mapbox style
- [ ] https://github.com/openlayers/ol-mapbox-style/issues/138 Interpret zoom level in the same way mapbox does
- [ ] https://github.com/openlayers/ol-mapbox-style/issues/139 Rendering of small features, min-width?

**Comment by pakb** [comment]
https://github.com/openlayers/ol-mapbox-style/issues/139 fixed 🎉

**Comment by davidoesch** [comment]
Can't wait for the demo link

**Comment by danduk82** [comment]
@swissgis3084, could you give a feedback on this codepen please: https://codepen.io/pakb/pen
/YgrYRK?editors=0010

@pakb please take the feedback of @swissgis3084 to continue the discussion with @ahocevar on issue [138](https://github.com/openlayers/ol-mapbox-style/issues/138)

For me it is quite acceptable, but I would like the author to criticize it constructively first (@swissgis3084), before we decide if we go the long way (option 2 in issue 138)

**Comment by pakb** [comment]
To me it starts to be pretty close to the rendering of mapbox (perfect match will be very hard to attain).
One last thing I might give a look into is font anti-aliasing. Seems like mapbox is smoothing fonts, but OL6 is not.

**Comment by ahocevar** [comment]
What exactly do you mean here @pakb?

Font smoothing is done by the Canvas API in OpenLayers. For performance reasons, text is pre-rendered to images, which are then placed on the layer's target canvas. During panning and zooming, we place text at subpixel locations to avoid jitter, at the cost of blurriness. When the map view is stable, we render it pixel-exact. For text along lines, we always have to use subpixel rendering because of the rotation.

**Comment by pakb** [comment]
@ahocevar see here on the left is OL rendering and on the right Mapbox

![image](https://user-images.githubusercontent.com/11540521/54215983-21ffed00-44e9-11e9-8963-1fb0b97206f2.png)

Could be simply a matter of font-color

**Comment by ahocevar** [comment]
Looks more like the text stroke is rendered above the fill in mapbox-gl, whereas it is rendered below the fill in OpenLayers. Or it's a font weight issue.

**Comment by swissgis3084** [comment]
> @swissgis3084, could you give a feedback on this codepen please: https://codepen.io/pakb/full/YgrYRK
> @pakb please take the feedback of @swissgis3084 to continue the discussion with @ahocevar on issue 138
> For me it is quite acceptable, but I would like the author to criticize it constructively first (@swissgis3084), before we decide if we go the long way (option 2 in issue 138)

can't really judge from codepen. it runs really slow. but from what i see the zoom levels seem to be pretty good

**Comment by pakb** [comment]
this is no longer relevant as we will be using Mapbox renderer (see https://github.com/geoadmin/mf-geoadmin3/pull/4925)
